mne/rank.py

```diff
@@ -65,27 +59,56 @@ def estimate_rank(data, tol='auto', return_singular=False, norm=True,
         return rank
 
 
+def _ratio_rank(s, ratio):
+    # Allow computing singular values as if we had been in float32
+    # precision, otherwise e.g. SSS->write->read->proj will create a
+    # larger shelf in the singular value curve
+    dtype = np.float32 if ratio == 'ratio32' else np.float64
+    min_ = np.finfo(dtype).eps * s[0]
+    ratios = -np.diff(np.log(np.maximum(s, min_)))
+    # Restrict it not to be in the first 10% of components. Should be
+    # reasonable for M/EEG data. Without this, we must set the ratio
+    # below much higher, which then is problematic for other data.
```

Comment:
What happens if the signal is full-rank? Would it then pick up component 10 as rank?

Reply:
Or is this captured by ``this_ratio`` below?

Reply:
See the issue description where I try to describe this scenario:

> 2. The data are actually full rank (the max-ratio becomes essentially random)
>
> The best way I have found to fix these (by trying many things to see what did and didn't pass in tests) was to limit the consideration of max rank ratio to ignore the first 10% of singular values, and to require a minimum ratio of at least 2., otherwise consider it full rank.

So the `this_ratio < 2` is meant to look for a real shelf. This value was chosen by seeing what worked and what didn't in the test suite.